### ISSUE TITLE
2.11/s3: Benchmarking seaweedfs/s3 via hsbench causes runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by ginkel** [chrislusf/seaweedfs#1622]
I have set up a SeaweedFS cluster with the following components:

3x master (`weed master -mdir=/data -volumePreallocate -volumeSizeLimitMB=1024 -ip=10.132.0.xx -mdir=/data -peers=10.132.0.58:9333,10.132.0.60:9333,10.132.0.57:9333`)
4x volume (`weed volume -dir=/data -max=0 -ip=10.132.0.xx -mserver=10.132.0.58:9333,10.132.0.60:9333,10.132.0.57:9333 -dataCenter=dc -rack=rack -dir=/data`)
4x filer+s3 (`weed filer -ip=10.132.0.xx -master=10.132.0.58:9333,10.132.0.60:9333,10.132.0.57:9333 -peers=10.132.0.58:8888,10.132.0.60:8888,10.132.0.57:8888,10.132.0.56:8888 -s3`)

The deployment is dockerized (using the `chrislusf/seaweedfs` Docker image), so please don't be confused about data being held in `/data` for all processes - these are actually disjunct.

The cluster is deployed on four GCE n1-standard-8 instances (8 vCPUs, 30 GB RAM) with local NVMe attached as storage, which is formatted with XFS and used to host the data of the volume container.

The filer is configured to use leveldb2 as filer store:

```
[leveldb2]
enabled = true
dir = "/data"
```

**Describe the bug**
When hitting the cluster with a synthetic benchmark using [hsbench](https://github.com/markhpc/hsbench) (see below for the benchmark script), the filer processes start reporting null-pointer de-references after a while:

```
2020/11/17 18:46:30 http: panic serving 10.132.0.59:47610: runtime error: invalid memory address or nil pointer dereference
goroutine 4136287 [running]:
net/http.(*conn).serve.func1(0xc00284cdc0)
        /home/travis/.gimme/versions/go1.15.5.linux.amd64/src/net/http/server.go:1801 +0x147
panic(0x1d4f6c0, 0x3228030)
        /home/travis/.gimme/versions/go1.15.5.linux.amd64/src/runtime/panic.go:975 +0x47a
github.com/chrislusf/seaweedfs/weed/s3api.(*S3ApiServer).proxyToFiler(0xc00048c5b0, 0x2452da0, 0xc00d9bef20, 0xc021cd6a00, 0xc03307a4e0, 0x5e, 0x212fae8)
        /home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/s3api/s3api_object_handlers.go:269 +0x89d
github.com/chrislusf/seaweedfs/weed/s3api.(*S3ApiServer).DeleteObjectHandler(0xc00048c5b0, 0x2452da0, 0xc00d9bef20, 0xc021cd6a00)
        /home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/s3api/s3api_object_handlers.go:125 +0x2eb
github.com/chrislusf/seaweedfs/weed/s3api.track.func1(0x245a5a0, 0xc02a0b2620, 0xc021cd6a00)
        /home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/s3api/stats.go:34 +0x23c
net/http.HandlerFunc.ServeHTTP(0xc00aa841a0, 0x245a5a0, 0xc02a0b2620, 0xc021cd6a00)
        /home/travis/.gimme/versions/go1.15.5.linux.amd64/src/net/http/server.go:2042 +0x44
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00a9ca0c0, 0x245a5a0, 0xc02a0b2620, 0xc01422f500)
        /home/travis/gopath/pkg/mod/github.com/gorilla/mux@v1.7.4/mux.go:210 +0xd3
net/http.serverHandler.ServeHTTP(0xc0006b82a0, 0x245a5a0, 0xc02a0b2620, 0xc01422f500)
        /home/travis/.gimme/versions/go1.15.5.linux.amd64/src/net/http/server.go:2843 +0xa3
net/http.(*conn).serve(0xc00284cdc0, 0x2464860, 0xc019359cc0)
        /home/travis/.gimme/versions/go1.15.5.linux.amd64/src/net/http/server.go:1925 +0x8ad
created by net/http.(*Server).Serve
        /home/travis/.gimme/versions/go1.15.5.linux.amd64/src/net/http/server.go:2969 +0x36c
```

**System Setup**
- Ubuntu 20.04 Linux instance-seaweedfs-0 5.4.0-1029-gcp #31-Ubuntu SMP Wed Oct 21 19:38:01 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- Docker 19.03.8
- output of `weed version`: `version 30GB 2.11 98827d6 linux amd64`
- if using filer, show the content of `filer.toml`

```
# A sample TOML config file for SeaweedFS filer store
# Used with "weed filer" or "weed server -filer"
# Put this file to one of the location, with descending priority
#    ./filer.toml
#    $HOME/.seaweedfs/filer.toml
#    /etc/seaweedfs/filer.toml

####################################################
# Customizable filer server options
####################################################
[filer.options]
# with http DELETE, by default the filer would check whether a folder is empty.
# recursive_delete will delete all sub folders and files, similar to "rm -Rf"
recursive_delete = false
# directories under this folder will be automatically creating a separate bucket
buckets_folder = "/buckets"
buckets_fsync = [          # a list of buckets with all write requests fsync=true
        "important_bucket",
        "should_always_fsync",
]

####################################################
# The following are filer store options
####################################################

[leveldb2]
# local on disk, mostly for simple single-machine setup, fairly scalable
# faster than previous leveldb, recommended.
enabled = true
dir = "/data"                                       # directory to store level db files

[mysql]  # or tidb
# CREATE TABLE IF NOT EXISTS filemeta (
#   dirhash     BIGINT         COMMENT 'first 64 bits of MD5 hash value of directory field',
#   name        VARCHAR(1000)  COMMENT 'directory or file name',
#   directory   TEXT           COMMENT 'full path to parent directory',
#   meta        LONGBLOB,
#   PRIMARY KEY (dirhash, name)
# ) DEFAULT CHARSET=utf8;

enabled = false
hostname = "localhost"
port = 3306
username = "root"
password = ""
database = ""              # create or use an existing database
connection_max_idle = 2
connection_max_open = 100
interpolateParams = false

[postgres] # or cockroachdb
# CREATE TABLE IF NOT EXISTS filemeta (
#   dirhash     BIGINT,
#   name        VARCHAR(65535),
#   directory   VARCHAR(65535),
#   meta        bytea,
#   PRIMARY KEY (dirhash, name)
# );
enabled = false
hostname = "localhost"
port = 5432
username = "postgres"
password = ""
database = ""              # create or use an existing database
sslmode = "disable"
connection_max_idle = 100
connection_max_open = 100

[cassandra]
# CREATE TABLE filemeta (
#    directory varchar,
#    name varchar,
#    meta blob,
#    PRIMARY KEY (directory, name)
# ) WITH CLUSTERING ORDER BY (name ASC);
enabled = false
keyspace="seaweedfs"
hosts=["10.132.0.60:9042", "10.132.0.57:9042", "10.132.0.56:9042"]
username=""
password=""

[redis2]
enabled = false
address  = "localhost:6379"
password = ""
database = 0

[redis_cluster2]
enabled = false
addresses = [
    "localhost:30001",
    "localhost:30002",
    "localhost:30003",
    "localhost:30004",
    "localhost:30005",
    "localhost:30006",
]
password = ""
# allows reads from slave servers or the master, but all writes still go to the master
readOnly = true
# automatically use the closest Redis server for reads
routeByLatency = true

[etcd]
enabled = false
servers = "10.132.0.60:2379,10.132.0.57:2379,10.132.0.56:2379"
timeout = "3s"

[mongodb]
enabled = false
uri = "mongodb://localhost:27017"
option_pool_size = 0
database = "seaweedfs"

[elastic7]
enabled = false
servers = [
    "http://localhost1:9200",
    "http://localhost2:9200",
    "http://localhost3:9200",
]
username = ""
password = ""
sniff_enabled = false
healthcheck_enabled = false
# increase the value is recommend, be sure the value in Elastic is greater or equal here
index.max_result_window = 10000
```

**Expected behavior**
The filer process does not crash.

**Screenshots**
n/a

**Benchmark script**
```
#!/bin/bash

COUNT="3"
MODES="cxiplgdcx"
OBJECT_SIZE="256B"
THREADS="300"

while [[ "$#" -gt 0 ]]; do
    case $1 in
        --count) COUNT="$2"; shift ;;
        --modes) MODES="$2"; shift ;;
        --object-size) OBJECT_SIZE="$2"; shift ;;
        --threads) THREADS="$2"; shift ;;
        *) echo "Unknown parameter passed: $1"; exit 1 ;;
    esac
    shift
done

COUNTER=0
while [  $COUNTER -lt $COUNT ]; do
    let COUNTER=COUNTER+1
    hsbench \
      -a accesstoken -s secret \
      -u http://instance-seaweedfs-0-int.domain \
      -m "$MODES" \
      -z "$OBJECT_SIZE" \
      -d "300" \
      -b "10" \
      -t "$THREADS" \
      -bp "hsbench-$COUNTER-1-$(date +%s)" \
      -o "hsbench-instance-seaweedfs-0-int.domain-$COUNTER.csv" \
      > "hsbench-instance-seaweedfs-0-int.domain-$COUNTER-summary.out" &

    hsbench \
      -a accesstoken -s secret \
      -u http://instance-seaweedfs-1-int.domain \
      -m "$MODES" \
      -z "$OBJECT_SIZE" \
      -d "300" \
      -b "10" \
      -t "$THREADS" \
      -bp "hsbench-$COUNTER-2-$(date +%s)" \
      -o "hsbench-instance-seaweedfs-1-int.domain-$COUNTER.csv" \
      > "hsbench-instance-seaweedfs-1-int.domain-$COUNTER-summary.out" &

    hsbench \
      -a accesstoken -s secret \
      -u http://instance-seaweedfs-2-int.domain \
      -m "$MODES" \
      -z "$OBJECT_SIZE" \
      -d "300" \
      -b "10" \
      -t "$THREADS" \
      -bp "hsbench-$COUNTER-3-$(date +%s)" \
      -o "hsbench-instance-seaweedfs-2-int.domain-$COUNTER.csv" \
      > "hsbench-instance-seaweedfs-2-int.domain-$COUNTER-summary.out" &

    hsbench \
      -a accesstoken -s secret \
      -u http://instance-seaweedfs-3-int.domain \
      -m "$MODES" \
      -z "$OBJECT_SIZE" \
      -d "300" \
      -b "10" \
      -t "$THREADS" \
      -bp "hsbench-$COUNTER-4-$(date +%s)" \
      -o "hsbench-instance-seaweedfs-3-int.domain-$COUNTER.csv" \
      > "hsbench-instance-seaweedfs-3-int.domain-$COUNTER-summary.out" &

done

echo "Waiting for jobs to finsh..."

wait
```

**Comment by chrislusf** [comment]
Added a fix for this nil error. You can get the updates in this image `docker pull chrislusf/seaweedfs:dev` now. 

btw: The image is usually 15 minutes after the code commits.

Thanks for the benchmark scripts. This would be very useful!

**Comment by chrislusf** [comment]
This benchmark script seems only hitting the s3 endpoints one by one instead of in parallel? Just to confirm.

**Comment by chrislusf** [comment]
Your test with 300 threads seems really high.

Here is the result from my local laptop, to an external SSD with NAND Flash & USB 3.1 Gen 2. 

```
benchmark$ hsbench -a accesstoken -s secret -z 4K -d 10 -t 10 -b 10 -u http://localhost:8333 -m "cxipgdx" -bp "hsbench-"
2020/11/17 12:40:22 Hotsauce S3 Benchmark Version 0.1
2020/11/17 12:40:22 Parameters:
2020/11/17 12:40:22 url=http://localhost:8333
2020/11/17 12:40:22 object_prefix=
2020/11/17 12:40:22 bucket_prefix=hsbench-
2020/11/17 12:40:22 region=us-east-1
2020/11/17 12:40:22 modes=cxipgdx
2020/11/17 12:40:22 output=
2020/11/17 12:40:22 json_output=
2020/11/17 12:40:22 max_keys=1000
2020/11/17 12:40:22 object_count=-1
2020/11/17 12:40:22 bucket_count=10
2020/11/17 12:40:22 duration=10
2020/11/17 12:40:22 threads=10
2020/11/17 12:40:22 loops=1
2020/11/17 12:40:22 size=4K
2020/11/17 12:40:22 interval=1.000000
2020/11/17 12:40:22 Running Loop 0 BUCKET CLEAR TEST
2020/11/17 12:40:22 Loop: 0, Int: TOTAL, Dur(s): 0.0, Mode: BCLR, Ops: 0, MB/s: 0.00, IO/s: 0, Lat(ms): [ min: 0.0, avg: 0.0, 99%: 0.0, max: 0.0 ], Slowdowns: 0
2020/11/17 12:40:22 Running Loop 0 BUCKET DELETE TEST
2020/11/17 12:40:22 Loop: 0, Int: TOTAL, Dur(s): 0.0, Mode: BDEL, Ops: 0, MB/s: 0.00, IO/s: 0, Lat(ms): [ min: 0.0, avg: 0.0, 99%: 0.0, max: 0.0 ], Slowdowns: 0
2020/11/17 12:40:22 Running Loop 0 BUCKET INIT TEST
2020/11/17 12:40:22 Loop: 0, Int: TOTAL, Dur(s): 0.0, Mode: BINIT, Ops: 10, MB/s: 0.00, IO/s: 3301, Lat(ms): [ min: 1.7, avg: 2.5, 99%: 2.9, max: 2.9 ], Slowdowns: 0
2020/11/17 12:40:22 Running Loop 0 OBJECT PUT TEST
2020/11/17 12:40:24 Loop: 0, Int: 0, Dur(s): 1.0, Mode: PUT, Ops: 2775, MB/s: 10.84, IO/s: 2775, Lat(ms): [ min: 1.4, avg: 3.5, 99%: 42.7, max: 114.4 ], Slowdowns: 0
2020/11/17 12:40:24 Loop: 0, Int: 1, Dur(s): 1.0, Mode: PUT, Ops: 3617, MB/s: 14.13, IO/s: 3617, Lat(ms): [ min: 1.4, avg: 2.8, 99%: 24.4, max: 87.9 ], Slowdowns: 0
2020/11/17 12:40:25 Loop: 0, Int: 2, Dur(s): 1.0, Mode: PUT, Ops: 3551, MB/s: 13.87, IO/s: 3551, Lat(ms): [ min: 1.3, avg: 2.8, 99%: 24.6, max: 58.4 ], Slowdowns: 0
2020/11/17 12:40:26 Loop: 0, Int: 3, Dur(s): 1.0, Mode: PUT, Ops: 4019, MB/s: 15.70, IO/s: 4019, Lat(ms): [ min: 1.3, avg: 2.5, 99%: 5.7, max: 96.0 ], Slowdowns: 0
2020/11/17 12:40:27 Loop: 0, Int: 4, Dur(s): 1.0, Mode: PUT, Ops: 3852, MB/s: 15.05, IO/s: 3852, Lat(ms): [ min: 1.4, avg: 2.6, 99%: 5.8, max: 54.3 ], Slowdowns: 0
2020/11/17 12:40:28 Loop: 0, Int: 5, Dur(s): 1.0, Mode: PUT, Ops: 3998, MB/s: 15.62, IO/s: 3998, Lat(ms): [ min: 1.4, avg: 2.5, 99%: 5.2, max: 60.0 ], Slowdowns: 0
2020/11/17 12:40:29 Loop: 0, Int: 6, Dur(s): 1.0, Mode: PUT, Ops: 3935, MB/s: 15.37, IO/s: 3935, Lat(ms): [ min: 1.3, avg: 2.4, 99%: 5.4, max: 50.5 ], Slowdowns: 0
2020/11/17 12:40:30 Loop: 0, Int: 7, Dur(s): 1.0, Mode: PUT, Ops: 3935, MB/s: 15.37, IO/s: 3935, Lat(ms): [ min: 1.4, avg: 2.6, 99%: 23.3, max: 47.6 ], Slowdowns: 0
2020/11/17 12:40:31 Loop: 0, Int: 8, Dur(s): 1.0, Mode: PUT, Ops: 3872, MB/s: 15.12, IO/s: 3872, Lat(ms): [ min: 1.3, avg: 2.6, 99%: 22.2, max: 48.9 ], Slowdowns: 0
2020/11/17 12:40:32 Loop: 0, Int: 9, Dur(s): 1.0, Mode: PUT, Ops: 3761, MB/s: 14.69, IO/s: 3761, Lat(ms): [ min: 1.4, avg: 2.7, 99%: 6.0, max: 52.6 ], Slowdowns: 0
2020/11/17 12:40:32 Loop: 0, Int: TOTAL, Dur(s): 10.0, Mode: PUT, Ops: 37325, MB/s: 14.58, IO/s: 3732, Lat(ms): [ min: 1.3, avg: 2.7, 99%: 22.3, max: 114.4 ], Slowdowns: 0
2020/11/17 12:40:32 Running Loop 0 OBJECT GET TEST
2020/11/17 12:40:33 Loop: 0, Int: 0, Dur(s): 1.0, Mode: GET, Ops: 4769, MB/s: 18.63, IO/s: 4769, Lat(ms): [ min: 1.0, avg: 2.0, 99%: 3.0, max: 105.2 ], Slowdowns: 0
2020/11/17 12:40:34 Loop: 0, Int: 1, Dur(s): 1.0, Mode: GET, Ops: 4999, MB/s: 19.53, IO/s: 4999, Lat(ms): [ min: 1.1, avg: 2.0, 99%: 3.0, max: 104.5 ], Slowdowns: 0
2020/11/17 12:40:35 Loop: 0, Int: 2, Dur(s): 1.0, Mode: GET, Ops: 4771, MB/s: 18.64, IO/s: 4771, Lat(ms): [ min: 1.1, avg: 2.1, 99%: 3.0, max: 48.1 ], Slowdowns: 0
2020/11/17 12:40:36 Loop: 0, Int: 3, Dur(s): 1.0, Mode: GET, Ops: 4790, MB/s: 18.71, IO/s: 4790, Lat(ms): [ min: 1.0, avg: 2.1, 99%: 3.1, max: 67.3 ], Slowdowns: 0
2020/11/17 12:40:37 Loop: 0, Int: 4, Dur(s): 1.0, Mode: GET, Ops: 4795, MB/s: 18.73, IO/s: 4795, Lat(ms): [ min: 1.1, avg: 2.0, 99%: 3.1, max: 29.7 ], Slowdowns: 0
2020/11/17 12:40:38 Loop: 0, Int: 5, Dur(s): 1.0, Mode: GET, Ops: 4593, MB/s: 17.94, IO/s: 4593, Lat(ms): [ min: 1.1, avg: 2.1, 99%: 3.2, max: 40.0 ], Slowdowns: 0
2020/11/17 12:40:39 Loop: 0, Int: 6, Dur(s): 1.0, Mode: GET, Ops: 4650, MB/s: 18.16, IO/s: 4650, Lat(ms): [ min: 1.1, avg: 2.1, 99%: 3.2, max: 29.1 ], Slowdowns: 0
2020/11/17 12:40:40 Loop: 0, Int: TOTAL, Dur(s): 7.8, Mode: GET, Ops: 37325, MB/s: 18.65, IO/s: 4775, Lat(ms): [ min: 1.0, avg: 2.1, 99%: 3.1, max: 105.2 ], Slowdowns: 0
2020/11/17 12:40:40 Running Loop 0 OBJECT DELETE TEST
2020/11/17 12:40:41 Loop: 0, Int: 0, Dur(s): 1.0, Mode: DEL, Ops: 302, MB/s: 1.18, IO/s: 302, Lat(ms): [ min: 20.8, avg: 32.7, 99%: 52.3, max: 54.6 ], Slowdowns: 0
2020/11/17 12:40:42 Loop: 0, Int: 1, Dur(s): 1.0, Mode: DEL, Ops: 307, MB/s: 1.20, IO/s: 307, Lat(ms): [ min: 21.4, avg: 32.2, 99%: 54.0, max: 55.8 ], Slowdowns: 0
2020/11/17 12:40:43 Loop: 0, Int: 2, Dur(s): 1.0, Mode: DEL, Ops: 311, MB/s: 1.21, IO/s: 311, Lat(ms): [ min: 19.8, avg: 32.2, 99%: 51.5, max: 53.3 ], Slowdowns: 0
2020/11/17 12:40:44 Loop: 0, Int: 3, Dur(s): 1.0, Mode: DEL, Ops: 309, MB/s: 1.21, IO/s: 309, Lat(ms): [ min: 20.5, avg: 32.7, 99%: 65.4, max: 68.2 ], Slowdowns: 0
2020/11/17 12:40:45 Loop: 0, Int: 4, Dur(s): 1.0, Mode: DEL, Ops: 310, MB/s: 1.21, IO/s: 310, Lat(ms): [ min: 21.7, avg: 32.3, 99%: 62.5, max: 65.0 ], Slowdowns: 0
2020/11/17 12:40:46 Loop: 0, Int: 5, Dur(s): 1.0, Mode: DEL, Ops: 302, MB/s: 1.18, IO/s: 302, Lat(ms): [ min: 22.4, avg: 32.8, 99%: 72.8, max: 74.9 ], Slowdowns: 0
2020/11/17 12:40:47 Loop: 0, Int: 6, Dur(s): 1.0, Mode: DEL, Ops: 300, MB/s: 1.17, IO/s: 300, Lat(ms): [ min: 23.5, avg: 33.4, 99%: 44.3, max: 48.5 ], Slowdowns: 0
2020/11/17 12:40:48 Loop: 0, Int: 7, Dur(s): 1.0, Mode: DEL, Ops: 295, MB/s: 1.15, IO/s: 295, Lat(ms): [ min: 23.7, avg: 33.9, 99%: 46.3, max: 60.4 ], Slowdowns: 0
2020/11/17 12:40:49 Loop: 0, Int: 8, Dur(s): 1.0, Mode: DEL, Ops: 299, MB/s: 1.17, IO/s: 299, Lat(ms): [ min: 20.5, avg: 33.4, 99%: 52.2, max: 54.2 ], Slowdowns: 0
2020/11/17 12:40:50 Loop: 0, Int: 9, Dur(s): 1.0, Mode: DEL, Ops: 304, MB/s: 1.19, IO/s: 304, Lat(ms): [ min: 22.7, avg: 32.5, 99%: 47.7, max: 50.4 ], Slowdowns: 0
2020/11/17 12:40:50 Loop: 0, Int: TOTAL, Dur(s): 10.0, Mode: DEL, Ops: 3049, MB/s: 1.19, IO/s: 305, Lat(ms): [ min: 19.8, avg: 32.8, 99%: 59.0, max: 74.9 ], Slowdowns: 0
2020/11/17 12:40:50 Running Loop 0 BUCKET DELETE TEST
2020/11/17 12:40:52 Loop: 0, Int: 0, Dur(s): 1.0, Mode: BDEL, Ops: 0, MB/s: 0.00, IO/s: 0, Lat(ms): [ min: 0.0, avg: 0.0, 99%: 0.0, max: 0.0 ], Slowdowns: 0
2020/11/17 12:40:52 Loop: 0, Int: TOTAL, Dur(s): 1.4, Mode: BDEL, Ops: 10, MB/s: 0.00, IO/s: 7, Lat(ms): [ min: 1216.6, avg: 1332.1, 99%: 1377.8, max: 1377.8 ], Slowdowns: 0
```

**Comment by ginkel** [comment]
Excellent, thanks a lot! I can no longer reproduce the error with the `dev` tag. There may still be problems for 4MiB file size with 30 concurrent requests, but I have not yet analyzed the error. After doing so I will probably open another ticket.
